feat(controller): Move deployments to KubeArmorController

DelusionalOptimist opened the pull request:
**Purpose of PR?**:
Fixes #1249 

**Does this PR introduce a breaking change?**
Not a breaking change but since deployments are being updated, updates would've to be carried accordingly.

**If the changes in this PR are manually verified, list down the scenarios covered:**:
* helm installation

**Additional information for reviewer?** :
* Deprecate KubearmorPolicy, KubearmorHostPolicy, KubearmorAnnotation controllers and use the single KubeArmorController instead.
* Update helm charts


**Checklist:**
- [ ] Bug fix. Fixes #<issue number>
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Commit has unit tests
- [ ] Commit has integration tests

<!--

The PR title message must follow convention:
`<type>(<scope>): <subject>`.

Where: <br />
- `type` is to define what type of PR is this.
  Most common types are:
    - `feat`      - for new features, not a new feature for build script
    - `fix`       - for bug fixes or improvements, not a fix for build script
    - `chore`     - changes not related to production code
    - `docs`      - changes related to documentation
    - `style`     - formatting, missing semi colons, linting fix etc; no significant production code changes
    - `test`      - adding missing tests, refactoring tests; no production code change
    - `refactor`  - refactoring production code, eg. renaming a variable or function name, there should not be any significant production code changes

- `scope` is a single word that best describes where the changes fit.
    - feature(`monitor`,`enforcer`)
    - test(`tests`, `bdd`)
    - chore(`build`)
- `subject` is a single line brief description of the changes made in the pull request.

-->

rksharma95 committed:
Update helm chart

Signed-off-by: rksharma95 <ramakant@accuknox.com>

DelusionalOptimist committed:
remove deprecated controllers

Signed-off-by: Rudraksh Pareek <rudraksh@accuknox.com>

kranurag7 reviewed (dismissed):
```diff
- GOBIN=$(PROJECT_DIR)/bin go install $(2) ;\
+ GOBIN=$(HOME)/go/bin go install $(2) ;\
```
The Makefile generated by kubebuilder is not installing controller-gen at `~/go/bin/controller-gen` rather it's installing the executable at `~/kubearmor/pkg/KubeArmorController/bin/controller-gen`. This fixes the CI locally and the images are building.

Requesting small changes here as well. 
https://github.com/kubearmor/KubeArmor/blob/main/contribution/development_guide.md#code-directories

DelusionalOptimist committed:
more chart updates

Signed-off-by: Rudraksh Pareek <rudraksh@accuknox.com>

DelusionalOptimist committed:
add helm chart publisher

Signed-off-by: Rudraksh Pareek <rudraksh@accuknox.com>

DelusionalOptimist committed:
Final helm chart updates. Changes introduced:
- Uniform naming convention in deployments
- Remove cluster-admin from chart & manifests (thanks to previous work
  by @kranurag7)
- Add a latest chart publishing workflow
- Minor changes to stable chart publisher workflow
- Minor chart updates based on review
- README updates

Signed-off-by: Rudraksh Pareek <rudraksh@accuknox.com>

kranurag7 reviewed (commented):
The changes look good to me on the helm side. 

Some minor comments. 
- Let's use `kubearmor/kubearmor-controller` in AWS and oke values file. I think that will work, we don't need a separate registry for each. 
- Should we use `genSelfSignedCert` instead of `genSignedCert` function while templating?

daemon1024 commented on .github/workflows/ci-test-helm-charts.yaml:
```diff
@@ -0,0 +1,85 @@
+      - name: Test KubeArmor using Ginkgo
+        run: |
+          go install -mod=mod github.com/onsi/ginkgo/v2/ginkgo
+          make -C tests/
```
Let's only run the smoke tests only here, since kubearmor will be tested in general anyway by other suites

DelusionalOptimist replied:
Running only the smoke test suite doesn't work : (

daemon1024 commented on KubeArmor/build/kubearmor-helm-test-values.yaml:
```diff
@@ -0,0 +1,28 @@
+    tag: latest
+  imagePullPolicy: Never
+  args:
+    - -logPath=/tmp/kubearmor.log
```
```suggestion
```

We don't need this imo

daemon1024 reviewed (changes requested):
LGTM in general, few nits inline

daemon1024 commented on .github/workflows/ci-latest-helm-chart-release.yaml:
```diff
@@ -0,0 +1,39 @@
+jobs:
+  publish-chart:
+    name: Update Stable Helm Chart With Latest Changes
+    if: ${{ (github.repository == 'delusionaloptimist/kubearmor') && (!contains(github.event.head_commit.message, '[Automated]')) }}
```
```suggestion
    if: ${{ (github.repository == 'kubearmor/kubearmor') && (!contains(github.event.head_commit.message, '[Automated]')) }}
```

Let's handle this before merging, Do we need to do any steps for helm repo setup for KubeArmor which I can help with?

DelusionalOptimist replied:
Good eye.

DelusionalOptimist replied:
@daemon1024 Can you create the `GH_ACCESS_TOKEN` secret? It'll be a PAT with read and write permissions to kubearmor repos

daemon1024 replied:
@DelusionalOptimist why do we need a PAT, why is GH Action token not enough?

DelusionalOptimist replied:
@daemon1024 AFAIK, the default GITHUB_TOKEN doesn't give enough permissions to push changes across repositories.

DelusionalOptimist replied:
@daemon1024 Here is a failed sample run using GH_TOKEN - https://github.com/DelusionalOptimist/KubeArmor/actions/runs/5335563256/jobs/9668964507#step:6:38

daemon1024 commented on deployments/get/defaults.go:
```diff
@@ -12,22 +12,21 @@ var port int32 = 32767
+	kubearmorClusterRoleBindingName           = "kubearmor-clusterrolebinding"
+	kubearmorClusterRoleName                  = "kubearmor-clusterrole"
+	relayServiceName                          = kubearmor
+	relayDeploymentName                       = "kubearmor-relay"
```
Let's keep them exported maybe will need to import them in kArmor

rksharma95 replied:
yes, can also be used in Operator.

daemon1024 commented on deployments/helm/Chart.yaml:
```diff
@@ -1,8 +1,8 @@
 apiVersion: v2
-appVersion: 1.16.0
+appVersion: v0.9
```
```suggestion
appVersion: v0.10.2
```
Or we can do this change when we update the stable release. WDYT?

DelusionalOptimist replied:
When we will merge this PR, the latest helm chart publisher would run. Though it would create a new tar file on `charts` repo, it won't create a commit updating the versions in chart.yaml in kubearmor repo. So even if we keep it to v0.9 or unchanged here, the packaged chart will take version from STABLE_RELEASE which is v0.9.

Also, we won't ever need to manually update chart and app version. This would be taken care of in the stable release workflow.

daemon1024 replied:
Then if ain't ever be updated, shouldn't we set it to something else so as it is not misleading or suggests it's unamanaged?

DelusionalOptimist replied:
Yeah I thought we could do it just this one time so that charts repo and the chart.yaml here seemed in sync. What else do you suggest :thinking:
